### PR TITLE
Remove performance regressions discovered in #1194

### PR DESCRIPTION
**Pull request by adamsitnik**
In #1194 I have analyzed and compared the results of all ML.NET end-to-end benchmarks for .NET Core 2.1 with native library for math and .NET Core 3.0 with managed CpuMath.

I found 3 quite huge regressions. After some extra profiling I found the reason:

**Most of the benchmarks use very short arrays, typically with just 4 or 14 floats. For such small arrays, the vectorization makes no sense.** This PR adds a very simple check: if the number of floats is small, don't run the vectorized code.  (/cc @AndyAyersMS this explains no perf boos in ML.NET scenarios after switch to new vectorized code)

|                             Type |              Method |     Toolchain |         Mean |      StdDev |
|--------------------------------- |-------------------- |-------------- |-------------:|------------:|
| KMeansAndLogisticRegressionBench |    TrainKMeansAndLR | netcoreapp2.1 |   566.601 ms | 168.8636 ms |
| KMeansAndLogisticRegressionBench |    TrainKMeansAndLR | netcoreapp3.0 |   568.968 ms |  92.8586 ms |
|            PredictionEngineBench | MakeIrisPredictions | netcoreapp2.1 |     3.896 ms |   0.0542 ms |
|            PredictionEngineBench | MakeIrisPredictions | netcoreapp3.0 |     3.820 ms |   0.0106 ms |

In 2.1 profiles we have calls to native dll:

image

Before my fix the code was calling the new managed CpuMath

image

Now there is no call to CpuMath if the vectors are small (and in most of our scenarios, they are). Moreover, the method gets inlined and we have no `pinvoke` cost compared to 2.1 and native dependency.

image

@danmosemsft I will need to take a look at the way ML.NET devides work into smaller tasks and change something on the upper level to take advantage of the vectorization.

### CODE REVIEW

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs [hunk 1]
```diff
@@ -259,19 +260,31 @@ public static void AddScale(float a, ReadOnlySpan<float> src, ReadOnlySpan<int>
...
-                {
-                    int index = indices[i];
-                    dst[index] += a * src[i];
-                }
+                AddScaleSimple(a, src, indices, dst, count);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void AddScaleSimple(float a, ReadOnlySpan<float> src, ReadOnlySpan<int> indices, Span<float> dst, int count)
```
Comment (eerhardt) [inline src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: (minor) It feels odd to use two different terms: `Vectorization` and `Simple` in this code. Can we stick with one term?  Maybe `AddScaleNonvectorized`?

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs [hunk 2]
```diff
@@ -259,19 +260,31 @@ public static void AddScale(float a, ReadOnlySpan<float> src, ReadOnlySpan<int>
 
             if (Avx.IsSupported)
             {
-                AvxIntrinsics.AddScaleSU(a, src, indices, dst, count);
+                if (VectorizationMakesSense(count))
```
Comment (eerhardt) [inline src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: We use a tertiary operator below, but we don't here. Is there are reason for the inconsistency?

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs [hunk 3]
```diff
@@ -247,6 +247,7 @@ public static void AddScale(float a, ReadOnlySpan<float> src, Span<float> dst, i
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (eerhardt) [inline src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: Why are we only marking some methods with `[MethodImpl(MethodImplOptions.AggressiveInlining)]` and not all?

Similarly, why are we checking `if VectorizationMakesSense` for only some methods, but not all?

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs [hunk 4]
```diff
@@ -761,5 +788,22 @@ public static void SdcaL1UpdateSparse(float primalUpdate, int count, ReadOnlySpa
                 }
             }
         }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool VectorizationMakesSense(int count)
+        {
+            if (Avx.IsSupported)
+            {
+                return count >= Vector256Alignment;
```
Comment (tannergooding) [inline src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: I don't think using `Vector256Alignment` here makes sense. We should have a separate named constant for this.

I also believe that the characteristics between AVX and SSE code will differ quite a bit.
* `SSE` should never come with any perf or startup penalty
  * Using SSE code requires at least 4 elements (16-bytes of data)
  * The worst case scenario is that the user has an array and the first 4 elements cross a cache-line or page boundary; but that should be a fairly rare occurence
* `AVX` (256-bit vectorized code, to be more precise), however, can cause processor downclocking
  * The Intel Optimization manual has more details on the downclocking that can occur, and we should find the "sweet" spot for each workload

I would imagine that all the numbers you have above are from AVX code, and that commenting out the AVX code-path and testing explicitly for just VEX-encoded SSE would show better numbers.

**Comment by tannergooding** [comment]
I would also like to know if the small array case we have for the benchmarks is representative of real-world scenarios. I would not expect that most ML scenarios deal with small amounts of data as inputs to these methods, and we may be adding additional unnecessary checks that will negatively impact the real-world scenarios (even if minor). If small array inputs are fairly representative, then there are likely some other changes we could make to the vectorized code that would otherwise improve this more.

**Comment by shauheen** [comment]
Ping @tannergooding @adamsitnik

**Comment by tannergooding** [comment]
@adamsitnik, I am still waiting for a response on https://github.com/dotnet/machinelearning/pull/1394#issuecomment-433970212

**Comment by adamsitnik** [comment]
> I would also like to know if the small array case we have for the benchmarks is representative of real-world scenarios.

@tannergooding I am currently studying and profiling ML.NET to find the answer. I started with reading the input files where we have some bottlenecks. Once I got the parsing done right I am going to investigate and optimize the way we split data into smaller jobs to take advantage of CpuMath.

**Comment by shauheen** [comment]
@adamsitnik do you think it would be possible to close this PR and continue discussion in an issue until you have your investigation completed? then you can update the branch and submit a new PR?

**Comment by adamsitnik** [comment]
@shauheen sure, good idea

**Pull request closed without merging**